text/0014-is-authorized-batch.md

```diff
@@ -0,0 +1,247 @@
+
+    let responses = authorizer.is_authorized(&requests, policy_set, entities);
+
+    let photos = vec![];
+
+    // we should have a request for each photo ID
+    // (in order) and a response for each request,
+    // so transitively we have a response for each
+    // photo ID.
+    for (photo_id, response) in photo_ids.iter().zip(responses.iter_mut()) {
+        match response.decision() {
+            Decision::Allow => {
+                let photo = ...;
+                photos.push(photo)
+            },
+            Decision::Deny => {
+                // skip
+            }
+        }
+    };
+
+    photos
+}
+```
+
+## Motivation
+
+At time of writing at least three language bindings exist
+([ref](https://github.com/cedar-policy/cedar-awesome#language-and-platform-integrations))
+that provide mechanisms to call the Cedar Authorizer's
+[`is_authorized` method](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L300).
+
+While the majority of requests likely operate on a single resource and can be
+handled efficiently by the current implementation, some requests (especially
+`list` requests) may need to check authorization against multiple resources, and
+it is inefficient to do a full round-trip for every request. Similarly, a
+higher-level operation may involve multiple lower-level actions on multiple
+resources.
+
+In naive comparison testing with the
+[k9securityio/cedar-py](https://github.com/k9securityio/cedar-py)
+`is_authorized` implementation against a
+[very naive `is_authorized_batch` implementation](https://github.com/glb/cedar-py/commit/98bf3e1fa2e90bb6f7326cbd846a3293fdbd960b),
+I observed speedups of at least 5x for even simple cases, as documented in
+[k9securityio/cedar-py#13](https://github.com/k9securityio/cedar-py/issues/13).
+
+@skuenzli also
+[reported](https://github.com/k9securityio/cedar-py/issues/13#issuecomment-1629790294)
+that:
+
+> The time observed in Python is consistently 2 orders of magnitude greater than
+> the authorizer.is_authorized call. These particular requests use a simple
+> policy, a simple schema, and ~170 entities.
+
+The initial thought was to simply create an `is_authorized_batch` API for the
+`cedar-py` Python binding, but in discussion in the Cedar Policy Language Slack
+there was support from @aaronjeline and @andrewmwells-amazon for exposing this
+API in Cedar core so that it could be picked up more easily by other language
+bindings, as they expect those bindings to have similar needs and performance
+concerns.
+
+## Detailed design
+
+### Accept a batch of requests with the current shape
+
+This approach creates a new `is_authorized_batch` function, similar to the
+`is_authorized` function, that accepts a list of `Request` structs
+([ref](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L2006))
+and returns a list of `Response` structs
```

Comment:
We'd want some way to associate which `Response` corresponds to which request right? We don't need that for the first example, but for listing queries we do.

Reply:
My original version of the example had

```rust
    for (request, response) in requests.iter().zip(responses.iter_mut()) {
```

but I realized that actually the requests weren't really what I wanted to correlate.

Reply:
+1 to andrew, I think we need some correlation but pairs seem good enough

Reply:
OK; not familiar with the syntax for this in Rust yet, will see what I can learn from `zip`'s implementation. I assume it would still be possible to zip with another iterable so that a person could do what I've shown in the list_photos example with correlating `photo_id` and `response`?

Reply:
I agree that we'd need to correlate request and response.  I think it should be the caller's responsibility to construct a single policy set, schema, and entities to support the requests.

re correlation options:

* correlating pairs works, but makes me nervous without some direct correlation mechanism even with JSON guaranteeing preservation of Array order, which is relevant for many/most binding libs
* perhaps request should support an optional `correlationId` property that is returned in the authz result
* I'd also be ok providing a map of requests keyed by a correlation id; I'm guessing I'll end up building that structure anyway so I can unpack the batch of results because authz result does not contain a copy of the request, only decision and diagnostics

Reply:
What I'm hearing is that folks are looking at:

```rust
pub fn is_authorized_batch(
    &self,
    requests: &Vec<Request>,
    policy_set: &PolicySet,
    entities: &Entities,
) -> Vec<Response> { ... }   
```

and thinking "it would be helpful to explicitly return the request with the response it created", something more like:

```rust
    pub fn is_authorized_batch(
        &self,
        requests: &Vec<Request>,
        policy_set: &PolicySet,
        entities: &Entities,
    ) -> Vec<(Request, Response)> { ... }
```

is that accurate? I'm also hearing an alternative suggest to extend `Request` to include a `correlationId` that would be copied to & returned back in the `Response`; not certain if that is seen as sufficient or whether there's a desire to have both the `correlationId` and the `Vec<(Request, Response)>` return in the final solution.

I'm contemplating the transport back over to the other side of the FFI boundary, which I know `cedar-py` uses JSON for, and the implied re-marshalling of the requests when sending them back alongside the responses. We don't have instrumentation results yet, but from the commentary and gut feel it seems like the JSON marshalling may be a material contributor to the latency and we might want to minimize the amount of data we're marshalling and passing across the boundary.

I'm not sure I understand the rationale for copying the request into the return value, as `responses[n]` is defined as the response for `requests[n]`, the caller will have `requests` already, and if they want to explicitly iterate through them together they can do so with `for (request, response) in requests.iter().zip(responses.iter_mut()) { ... }` as I outlined above and in the listing example.

Now that folks have had a couple of days for the comments to percolate, any new thoughts? Have I misunderstood the comments?